Modify S5527: Learn-As-You-Code migration (#2269)
SonarSource/rspec: rules/S5527/common/fix/validation.adoc

To fix the vulnerability of disabled hostname validation, it is strongly
recommended to first re-enable the default validation and fix the root cause: the validity of the certificate.

==== Use valid certificates

If a hostname validation failure prevents connecting to the target server, keep
in mind that **one system's code should not work around another system's problems**,

> as this creates unnecessary dependencies and can lead to reliability issues. +
> Therefore, the first solution is to change the remote host's certificate to
> match its identity. If the remote host is not under your control, consider replicating its
> service to a server whose certificate you can change yourself.
> 
> 
> In case the contacted host is located on a development machine, and if there
> is no other choice, try following this solution:

Author: ```suggestion
as this creates unnecessary dependencies and can lead to reliability issues.

Therefore, the first solution is to change the remote host's certificate to
match its identity. If the remote host is not under your control, consider replicating its
service to a server whose certificate you can change yourself.

In case the contacted host is located on a development machine, and if there
is no other choice, try following this solution:
```

To fix the vulnerability of disabled hostname validation, it is strongly
recommended to first re-enable the default validation.

To avoid running into problems with invalid hostnames, consider the following
sections.

==== Use valid certificates

If a hostname validation failure prevents connecting to the target server, keep
in mind that **one system's code should not work around another system's problems**,
as this creates unnecessary dependencies and can lead to reliability issues. +
Therefore, the first solution is to change the remote host's certificate to
match its identity. If the remote host is not under your control, consider replicating its
service to a server whose certificate you can change yourself.

In case the contacted host is located on a development machine, and if there
is no other choice, try following this solution:

* Create a self-signed certificate for that machine.
* Add this self-signed certificate to the system's trust store.

> * If the hostname is not `localhost`, set the correct host name in the `/etc/hosts` file.

Author: ```suggestion
* If the hostname is not `localhost`, set the correct host name in the `/etc/hosts` file.

```

To fix the vulnerability of disabled hostname validation, it is strongly

> recommended to first re-enable the default validation.
> 
> To avoid running into problems with invalid hostnames, consider the following
> sections.

Author: ```suggestion
recommended to first re-enable the default validation and fix the root cause: the validity of the certificate.
```

==== Use valid certificates

If a hostname validation failure prevents connecting to the target server, keep
in mind that **one system's code should not work around another system's problems**,
as this creates unnecessary dependencies and can lead to reliability issues. +
Therefore, the first solution is to change the remote host's certificate to
match its identity. If the remote host is not under your control, consider replicating its
service to a server whose certificate you can change yourself.

In case the contacted host is located on a development machine, and if there
is no other choice, try following this solution:

* Create a self-signed certificate for that machine.
* Add this self-signed certificate to the system's trust store.

> * If the hostname is not `localhost`, set the correct host name in the `/etc/hosts` file.

Author: ```suggestion
* If the hostname is not `localhost`, add the hostname in the `/etc/hosts` file.
```

sections.

==== Use valid certificates

If hostname validation failure prevents connection to the target server, keep
in mind that **one system's code should not work around another system's problems**,
as this creates unnecessary dependencies and can lead to reliability issues. +
Therefore, the first solution is to change the remote host's certificate to
match its identity. If the remote host is not under your control, replicate its
service to a server whose certificate you can change yourself.

If the remote hostname is dynamic, setting up on-the-fly certificate creation
thanks to scripting is a solution. +
If the remote hostname can only be distributed with an private DNS, unreachable
from outside, an alternative could be an intermediary reverse proxy located in
the internal network, with firewall protections that would set up a TLS proxy
thanks to the DNS access, while itself setting up a TLS certificate to fully
protect messages.

In case the contacted host is located on a development machine, this quick
solution is recommended:

* Create a self-signed certificate for that machine.
* Add this self-signed certificate to the system's trust store.

> * Set the correct host name in the `/etc/hosts` file.

Reviewer: I don't think the readers would understand what they are supposed to do here and it is probably not even necessary. Since you suggest using a self-signed certificate, they can just use localhost.

Author: I just changed it. Do you think it still make sense or would you still delete it ?

> * Set the correct host name in the `/etc/hosts` file.

Author: ```suggestion
* If the hostname is not `localhost`, set the correct host name in the `/etc/hosts` file.
```
